File: template/python3/SlideMin.py
from collections import deque

class SlideMin:
    """
    スライド最小値\n
    各区間[i, i+leng)の最小値を取得する：O(n)\n
    self.que = [(i1, v1), (i2, v2), ..., (ik, vk)]\n
    i1 < ... < ik and v1 < ... < vk
    """
    def __init__(self, leng):
        self.que = deque([])
        self.leng = leng

    def top(self):
        return self.que[0]

    def getQue(self):
        return self.que

    def update(self, index, value):
        while self.que:
            if self.que[-1][1] >= value:
                self.que.pop()
            else:
                break

        self.que.append((index, value))

        if self.que[0][0] == index - self.leng:
            self.que.popleft()

File: template/c++/SlideMin.cpp
#include <bits/stdc++.h>
using namespace std;

template<class T = int, class U = int>
class SlideMin {
    struct Pair {
        T index;
        U value;
        Pair(){};
        Pair(T index, U value): index(index), value(value) {};
    };

    private:
        deque<Pair> que;
        T length;
    public:
        SlideMin(T length) {
            this->length = length;
        }

        Pair top() {
            return this->que.front();
        }

        void update(T index, U value) {
            while (!this->que.empty() && this->que.back().value >= value) {
                this->que.pop_back();
            }
            this->que.emplace_back(index, value);

            if(this->que.front().index == index - this->length) this->que.pop_front();
        }
};
